Bluetooth: host: Fix RPA timeout expiration

joerchan opened the pull request:
Fix RPA timeout expiration when BT_EXT_ADV has been enabled.

Always invalidate the device RPA on RPA timeout.
This RPA is used by the scan roles, and the advertiser that was started
using the bt_le_adv_start API.

Call the RPA expire callback only on advertising sets that are enabled
and not being limited by a timeout and not using the identity address.
On RPA timeout always invalidate the RPA of advertising sets that are
disabled.

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/51208
Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/51228
Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/51247

Add shell support for RPA expire callback.
This prints a message when the callback is called on the specific
advertising set.
Add a command to control the return value of this callback for a
specific advertising set.

joerchan committed:
Revert "Bluetooth: Host: Fix resolvable address update with Ext Adv enabled"

This reverts commit 6c40427931dba65092d0eb050a2ed7e9e8069845.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: shell: Add shell support for RPA expire callback and control

Add shell support for RPA expire callback.
This prints a message when the callback is called on the specific
advertising set.
Add a command to control the return value of this callback for a
specific advertising set.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: host: Fix RPA timeout expiration

Fix RPA timeout expiration when BT_EXT_ADV has been enabled.

Always invalidate the device RPA on RPA timeout.
This RPA is used by the scan roles, and the advertiser that was started
using the bt_le_adv_start API.

Call the RPA expire callback only on advertising sets that are enabled
and not being limited by a timeout and not using the identity address.
On RPA timeout always invalidate the RPA of advertising sets that are
disabled.

Fixes: #51208
Fixes: #51228
Fixes: #51247

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: shell: Use a common help text for commands args on and off

Use a common help text for commands args on and off.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

cvinayak commented on subsys/bluetooth/host/id.c:
```diff
@@ -184,9 +184,10 @@ int bt_id_set_adv_random_addr(struct bt_le_ext_adv *adv,
+static void adv_disabled_rpa_invalidate(struct bt_le_ext_adv *adv, void *data)
 {
-	if (!atomic_test_bit(adv->flags, BT_ADV_LIMITED)) {
+	/* Don't invalide RPA of enabled advertising set. */
```
```suggestion
	/* Don't invalidate RPA of enabled advertising set. */
```

cvinayak commented on subsys/bluetooth/host/id.c:
```diff
@@ -410,25 +406,27 @@ static void adv_disable_rpa(struct bt_le_ext_adv *adv, void *data)
+		if (rpa_invalid) {
+			atomic_clear_bit(adv->flags, BT_ADV_RPA_VALID);
+		} else {
+			/* Submit the timeout in case no advertising set updates the their RPA
+			 * in the current period. This makes sure that the RPA timer is running.
+			 */
```
```suggestion
			/* Submit the timeout in case no advertising set updates their RPA
			 * in the current period. This makes sure that the RPA timer is running.
			 */
```

cvinayak reviewed (changes requested):
Minor typos.

kapi-no commented on subsys/bluetooth/host/id.c:
```diff
@@ -199,8 +200,13 @@ static void le_rpa_invalidate(void)
 		atomic_clear_bit(bt_dev.flags, BT_DEV_RPA_VALID);
 	}
 
+	/* Invalidate RPA of advertising sets that are disabled.
+	 *
+	 * Enabled advertising set have to call rpa_expired callback first to
+	 * determine if the RPA of the advertising set should be rotated.
+	 */
```
Please consider my comment:

https://github.com/zephyrproject-rtos/zephyr/issues/51247#issuecomment-1283705406

kapi-no replied:
@Thalley, @cvinayak and other stakeholders/reviewers of the ```rpa_expired``` callback feature. I suggested the potential modification in the linked comment above. We discussed it briefly with @joerchan. Please speak up if you have any counter-arguments/feedback regarding the conditions in which this callback is called.

Thalley replied:
I am not sure I follow the issue mentioned in that comment well enough to provide any additional comments at this point

carlescufi replied:
@jhedberg @jori-nordic @alwa-nordic @cvinayak we need input to resolve this.

kapi-no replied:
> Would it be useful for the application to get a new_rpa event? Right now rpa_expired is suggested as the event the application should use to synchronize with RPA changes. But that does not quite fit with the stopped-advertising-set case (unless we do assign new RPAs to stopped advertising sets), since a rotation would also happen on adv-start. Hopefully most applications are fine with just the 'expired' event.

I think the new_rpa event is unnecessary if the API user is familiar with Zephyr Bluetooth stack mechanics (we might improve the documentation for this reason). If the advertising is ongoing, the ```rpa_expired``` is also a signal that the new RPA will be used after the callback execution. If the advertising is stopped and the advertising set is not deleted, the ```rpa_expired``` callback should also notify the user about the RPA expiry. The API user should be aware at this point that the Bluetooth stack uses the new RPA upon the advertising restart.

Furthermore, the API user has full control over the RPA for its advertising set regardless of the advertising state (running / disabled). They can suspend the RPA rotation with the ```rpa_expired``` callback for an indefinite time while switching between advertising states. They can also enforce the rotation asynchronously with the ```bt_le_ext_adv_oob_get_local``` API. They have also the possibility to reject the RPA rotation in case of the asynchronous ```bt_le_ext_adv_oob_get_local``` API call, triggered by the user of some other advertising set (mitigates dependency on other advertising sets).

carlescufi replied:
Summary from @kapi-no:
The `rpa_expired` callback should be called every time when the `BT_ADV_RPA_VALID` flag is cleared for consistency (clearing causes RPA regeneration).

carlescufi commented:
@joerchan please update this PR with the [conclusion](https://github.com/zephyrproject-rtos/zephyr/pull/51258#discussion_r1005556543) in the comments.

joerchan committed:
Bluetooth: host: Call RPA expired on disabled advertising sets

Call the RPA expired callback on advertising sets even when they are
not enabled.
The RPA expired callback will now be called on advertising sets
when the BT_ADV_RPA_VALID flag is cleared for the set.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan commented:
I've updated the PR with the RPA expired callback now always being called when the RPA of the advertising set is no longer valid.
The application can choose to not invalide RPA on inactive advertising sets.

I've included this as a separate commit. I can squash this with the fix commit if that is preferred.

kapi-no commented:
@joerchan, now it seems that the ```rpa_expired``` callback works reliably. The ```BT_ADV_RPA_VALID``` flag clearing operation seems always to be preceded by execution of the ```rpa_expired``` callback (when the BT_EXT_ADV and BT_PRIVACY are enabled). Thanks for improving callback behaviour!

And once again sorry for introducing unwanted side effects for the Bluetooth Central privacy feature with my initial ```rpa_expired``` PR

kruithofa commented:
@jhedberg @alwa-nordic please re-review

cvinayak reviewed (approved):
quick pass.
